File: common/src/main/java/juuxel/adorn/item/DrinkInMugItem.java
package juuxel.adorn.item;

import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.ItemUseAnimation;
import net.minecraft.world.level.Level;

public class DrinkInMugItem extends Item {
    public DrinkInMugItem(Properties settings) {
        super(settings);
    }

    @Override
    public ItemStack finishUsingItem(ItemStack stack, Level world, LivingEntity user) {
        var result = super.finishUsingItem(stack, world, user);
        return user instanceof Player player && player.getAbilities().instabuild ? result : new ItemStack(AdornItems.MUG.get());
    }

    @Override
    public ItemUseAnimation getUseAnimation(ItemStack stack) {
        return ItemUseAnimation.DRINK;
    }
}

File: common/src/main/kotlin/juuxel/adorn/item/DrinkInMugItem.kt
package juuxel.adorn.item

import net.minecraft.entity.LivingEntity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.ItemStack
import net.minecraft.sound.SoundEvent
import net.minecraft.util.UseAction
import net.minecraft.world.World

open class DrinkInMugItem(settings: Settings) : ItemWithDescription(settings) {
    override fun finishUsing(stack: ItemStack, world: World, user: LivingEntity): ItemStack {
        val result = super.finishUsing(stack, world, user)
        return if (user is PlayerEntity && user.abilities.creativeMode) result else ItemStack(AdornItems.MUG)
    }

    override fun getUseAction(stack: ItemStack): UseAction =
        UseAction.DRINK

    override fun getEatSound(): SoundEvent = drinkSound
}
